### PR TITLE
extended tracing setup by elasticsearch instance

### PR DESCRIPTION
**Pull request by frzifus**
Signed-off-by: Benedikt Bongartz <bongartz@klimlive.de>

based on the standalone ES tested here: https://github.com/jaegertracing/jaeger-operator/pull/1937#issue-1261500680

Used for testing.
```bash
oc-4.10 process -f  resources/services/observatorium-traces-template.yaml -p NAMESPACE=observatorium-traces-testing  | oc-4.10 create -f -
```

Blocked by: https://github.com/rhobs/configuration/pull/242

---

## Update

We need to wait until Red Hat OpenShift distributed tracing platform 2.4 is released. Its scheduled for today (16.06.2022)

### CODE REVIEW

File: services/observatorium-traces.libsonnet [hunk 1]
```diff
@@ -2,6 +2,48 @@ local tracing = (import 'github.com/observatorium/observatorium/configuration/co
...
+          },
+          requests: {
+            cpu: '${ELASTICSEARCH_REQUEST_CPU}',
+            memory: '${ELASTICSEARCH_REQUEST_MEMORY}',
+          },
+        },
+      },
+      nodes: [
+        {
+          nodeCount: 1,
```
Comment (periklis) [inline services/observatorium-traces.libsonnet:32]: Is this maybe a goof candidate for a template parameter? I.e. do we want run single node ES on production too?

Comment (frzifus) [reply services/observatorium-traces.libsonnet:32]: indeed, but scaling only the nodes is not useful, as roles and redundancy policies can differ.
In consensus with @pavolloffay, we decided to keep it non-configurable for now and revisit this point later.

Comment (periklis) [reply services/observatorium-traces.libsonnet:32]: @pavolloffay I can certainly speak from my experience with AppSRE's onboard procedures that this is going to raise a blocker. Anything running on AppSRE's production requires at minimum an HA setup.

Comment (frzifus) [reply services/observatorium-traces.libsonnet:32]: thanks @periklis

Comment (frzifus) [reply services/observatorium-traces.libsonnet]: added some way to configure node types and count in e4d2009144797c375dd1073d58d612089b42e727. Iam not familiar with configuring ES and would appreciate feedback. (didnt test it yet, still waiting for the RHOSDT 2.4 release)

Comment (frzifus) [reply services/observatorium-traces.libsonnet]: Iam not an ES expert, but it seems to work with the split configuration.

Here is what i tested.
```bash
oc-4.10 process -f  resources/services/observatorium-traces-template.yaml -p NAMESPACE=observatorium-traces-testing -p ELASTICSEARCH_NODES_ALLINONE=0 -p ELASTICSEARCH_NODES_MASTER=1 -p ELASTICSEARCH_NODES_CLIENT=1 -p ELASTICSEARCH_NODES_DATA=1  | oc-4.10 apply -f -
```

And what i got:
```bash
NAME                                                              READY   STATUS             RESTARTS      AGE
shared-es-c-e71w31o2-0                                            2/2     Running            0             96s
shared-es-d-0cizwmwr-1-5cd6b86759-nxsb4                           2/2     Running            0             96s
shared-es-m-fk21lbua-0                                            2/2     Running            0             96s
```

File: services/observatorium-traces.libsonnet [hunk 2]
```diff
@@ -2,6 +2,75 @@ local tracing = (import 'github.com/observatorium/observatorium/configuration/co
...
+           proxyResources: {},
+           resources: {},
+           roles: [
+             'master',
+             'client',
+             'data',
+           ],
+           storage: {},
+         } else {}),
+        (if '${{ELASTICSEARCH_NODES_MASTER}}' != 0 then {
```
[the remark this answers is not in the document]
Comment (frzifus) [reply services/observatorium-traces.libsonnet:42]: We can mix all parameters.
```
$ oc-4.10 process -f  resources/services/observatorium-traces-template.yaml -p NAMESPACE=observatorium-traces-testing -p ELASTICSEARCH_NODES_ALLINONE=1 -p ELASTICSEARCH_NODES_MASTER=1 -p ELASTICSEARCH_NODES_CLIENT=1 -p ELASTICSEARCH_NODES_DATA=1 -p ELASTICSEARCH_NAME=test123  | oc-4.10 apply -f -

# got
NAME                                                        READY   STATUS    RESTARTS      AGE
observatorium-jaeger-rhobs-collector-b94489cd4-fdw5n        1/1     Running   2 (37s ago)   84s
observatorium-jaeger-rhobs-query-6688cd477b-fmxsd           3/3     Running   2 (38s ago)   84s
observatorium-jaeger-telemeter-collector-7ff4644f44-jdnww   1/1     Running   3 (32s ago)   81s
observatorium-jaeger-telemeter-query-554d6f867-7rszz        3/3     Running   3 (34s ago)   81s
observatorium-otel-collector-58df976b46-b4qb4               1/1     Running   0             91s
test123-c-wu1tef7d-0                                        2/2     Running   0             60s
test123-cdm-34nbdotj-1-69d45d87c-pl2k6                      2/2     Running   0             61s
test123-d-c59nlmoj-1-8568bf964c-9k8h8                       2/2     Running   0             60s
test123-m-xvfbuwa8-0                                        2/2     Running   0             60s

File: services/observatorium-traces.libsonnet (hunk 1 above)
Comment (periklis) [reply services/observatorium-traces.libsonnet]: In addition it makes using the right master count easy: https://github.com/openshift/elasticsearch-operator/blob/master/internal/elasticsearch/defaults.go#L24:2

File: resources/services/observatorium-traces-template.yaml [hunk 3]
```diff
@@ -3,6 +3,34 @@ kind: Template
...
+    namespace: ${NAMESPACE}
+  spec:
+    managementState: Managed
+    nodeSpec:
+      resources:
+        limits:
+          memory: ${ELASTICSEARCH_LIMIT_MEMORY}
+        requests:
+          cpu: ${ELASTICSEARCH_REQUEST_CPU}
+          memory: ${ELASTICSEARCH_REQUEST_MEMORY}
```
Comment (pavolloffay) [inline resources/services/observatorium-traces-template.yaml:23]: https://docs.openshift.com/container-platform/4.10/logging/config/cluster-logging-log-store.html says to use the same value for request and limit memory. So perhaps we can have a single variable for now and use it for request and limits.

File: resources/services/observatorium-traces-template.yaml [hunk 4]
```diff
@@ -3,6 +3,34 @@ kind: Template
...
+    nodes:
+    - nodeCount: ${{ELASTICSEARCH_NODE_COUNT}}
```
Comment (pavolloffay) [inline resources/services/observatorium-traces-template.yaml:25]: not sure if `{{}}` is correct

**Review by pavolloffay — APPROVED** [review]

**Comment by frzifus** [comment]
@periklis what do you think?

File: resources/services/observatorium-traces-template.yaml [hunk 5]
```diff
@@ -14,7 +42,16 @@ objects:
...
   spec:
-    strategy: allinone
+    storage:
+      elasticsearch:
+        doNotProvision: true
+        name: ${ELASTICSEARCH_NAME}
+        useCertManagement: true
+      options:
+        es:
+          index-prefix: rhobs
```
Comment (periklis) [inline resources/services/observatorium-traces-template.yaml:52]: Are these extra index prefixes working with our present [OpenDistro Roles](https://github.com/openshift/origin-aggregated-logging/blob/master/elasticsearch/sgconfig/roles.yml) for Jaeger?

**Review by periklis — COMMENTED** [review]
Looks good to me! Only double-checking on the index-prefixes.

File: resources/services/observatorium-traces-template.yaml (hunk 5 above)
Comment (frzifus) [reply resources/services/observatorium-traces-template.yaml:52]: the roles are independent of the index-prefix, right?
As far as i understood, all instances gets its own [curator and jaeger role certs](https://github.com/rhobs/configuration/pull/241/files/a4ac18eef353fe0ff2d1b418cbe4d549583a4fb0#diff-823e69dca2aebc3ebf77650ddbba4c2968224c3079b6eef5804bc70098395774R11).

Comment (periklis) [reply resources/services/observatorium-traces-template.yaml:52]: The roles have per se nothing to do with the certs. I am talking about authorization on the OpenDistro level. AFAICS your Jaeger definitions are `*jaeger-..*`, thus if you prefix your indices with `rhobs-jaeger...` you should be fine.

**Review by periklis — APPROVED** [review]

**Pull request merged**